// @zendeskgarden/container-scrollregion/dist/index.cjs.js
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var debounce = require('lodash.debounce');
var PropTypes = require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var debounce__default = /*#__PURE__*/_interopDefaultLegacy(debounce);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

function useScrollRegion(_ref) {
  var containerRef = _ref.containerRef,
      dependency = _ref.dependency;
  var _useState = React.useState(),
      containerTabIndex = _useState[0],
      setContainerTabIndex = _useState[1];
  var updateContainerTabIndex = React.useMemo(function () {
    return debounce__default["default"](function () {
      if (containerRef.current) {
        var regionContent = containerRef.current.children[0];
        var regionContentHeight = regionContent.scrollHeight;
        var regionContentWidth = regionContent.scrollWidth;
        var containerHeight = containerRef.current.offsetHeight;
        var containerWidth = containerRef.current.offsetWidth;
        if (regionContentWidth > containerWidth || regionContentHeight > containerHeight) {
          setContainerTabIndex(0);
        } else {
          setContainerTabIndex(undefined);
        }
      }
    }, 100);
  }, [containerRef, setContainerTabIndex]);
  React.useEffect(function () {
    addEventListener('resize', updateContainerTabIndex);
    updateContainerTabIndex();
    return function () {
      removeEventListener('resize', updateContainerTabIndex);
      updateContainerTabIndex.cancel();
    };
  }, [updateContainerTabIndex, dependency]);
  return containerTabIndex;
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

var _excluded = ["children", "render"];
var ScrollRegionContainer = function ScrollRegionContainer(_ref) {
  var children = _ref.children,
      _ref$render = _ref.render,
      render = _ref$render === void 0 ? children : _ref$render,
      props = _objectWithoutPropertiesLoose(_ref, _excluded);
  return React__default["default"].createElement(React__default["default"].Fragment, null, render(useScrollRegion(props)));
};
ScrollRegionContainer.propTypes = {
  children: PropTypes__default["default"].func,
  render: PropTypes__default["default"].func,
  containerRef: PropTypes__default["default"].any.isRequired,
  dependency: PropTypes__default["default"].any
};

exports.ScrollRegionContainer = ScrollRegionContainer;
exports.useScrollRegion = useScrollRegion;
