// @zendeskgarden/scripts/dist/index.js
"use strict";
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.netlifyToken = exports.netlifySiteId = exports.netlifyDeploy = exports.netlifyBandwidth = exports.lernaNew = exports.lernaChangelog = exports.githubToken = exports.githubRepository = exports.githubRelease = exports.githubPages = exports.githubDeploy = exports.githubCommit = exports.githubBranch = exports.cmdDu = exports.cmdDocgen = void 0;
var cmd_1 = require("./cmd");
Object.defineProperty(exports, "cmdDocgen", { enumerable: true, get: function () { return cmd_1.docgen; } });
Object.defineProperty(exports, "cmdDu", { enumerable: true, get: function () { return cmd_1.du; } });
var github_1 = require("./github");
Object.defineProperty(exports, "githubBranch", { enumerable: true, get: function () { return github_1.branch; } });
Object.defineProperty(exports, "githubCommit", { enumerable: true, get: function () { return github_1.commit; } });
Object.defineProperty(exports, "githubDeploy", { enumerable: true, get: function () { return github_1.deploy; } });
Object.defineProperty(exports, "githubPages", { enumerable: true, get: function () { return github_1.pages; } });
Object.defineProperty(exports, "githubRelease", { enumerable: true, get: function () { return github_1.release; } });
Object.defineProperty(exports, "githubRepository", { enumerable: true, get: function () { return github_1.repository; } });
Object.defineProperty(exports, "githubToken", { enumerable: true, get: function () { return github_1.token; } });
var lerna_1 = require("./lerna");
Object.defineProperty(exports, "lernaChangelog", { enumerable: true, get: function () { return lerna_1.changelog; } });
Object.defineProperty(exports, "lernaNew", { enumerable: true, get: function () { return lerna_1.new; } });
var netlify_1 = require("./netlify");
Object.defineProperty(exports, "netlifyBandwidth", { enumerable: true, get: function () { return netlify_1.bandwidth; } });
Object.defineProperty(exports, "netlifyDeploy", { enumerable: true, get: function () { return netlify_1.deploy; } });
Object.defineProperty(exports, "netlifySiteId", { enumerable: true, get: function () { return netlify_1.siteId; } });
Object.defineProperty(exports, "netlifyToken", { enumerable: true, get: function () { return netlify_1.token; } });
