// @zendeskgarden/scripts/dist/github/index.js
"use strict";
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.token = exports.tokenCommand = exports.repository = exports.repositoryCommand = exports.release = exports.releaseCommand = exports.pages = exports.pagesCommand = exports.deploy = exports.deployCommand = exports.commit = exports.commitCommand = exports.branch = exports.branchCommand = void 0;
var branch_1 = require("./branch");
Object.defineProperty(exports, "branchCommand", { enumerable: true, get: function () { return __importDefault(branch_1).default; } });
Object.defineProperty(exports, "branch", { enumerable: true, get: function () { return branch_1.execute; } });
var commit_1 = require("./commit");
Object.defineProperty(exports, "commitCommand", { enumerable: true, get: function () { return __importDefault(commit_1).default; } });
Object.defineProperty(exports, "commit", { enumerable: true, get: function () { return commit_1.execute; } });
var deploy_1 = require("./deploy");
Object.defineProperty(exports, "deployCommand", { enumerable: true, get: function () { return __importDefault(deploy_1).default; } });
Object.defineProperty(exports, "deploy", { enumerable: true, get: function () { return deploy_1.execute; } });
var pages_1 = require("./pages");
Object.defineProperty(exports, "pagesCommand", { enumerable: true, get: function () { return __importDefault(pages_1).default; } });
Object.defineProperty(exports, "pages", { enumerable: true, get: function () { return pages_1.execute; } });
var release_1 = require("./release");
Object.defineProperty(exports, "releaseCommand", { enumerable: true, get: function () { return __importDefault(release_1).default; } });
Object.defineProperty(exports, "release", { enumerable: true, get: function () { return release_1.execute; } });
var repository_1 = require("./repository");
Object.defineProperty(exports, "repositoryCommand", { enumerable: true, get: function () { return __importDefault(repository_1).default; } });
Object.defineProperty(exports, "repository", { enumerable: true, get: function () { return repository_1.execute; } });
var token_1 = require("./token");
Object.defineProperty(exports, "tokenCommand", { enumerable: true, get: function () { return __importDefault(token_1).default; } });
Object.defineProperty(exports, "token", { enumerable: true, get: function () { return token_1.execute; } });
