// @zendeskgarden/scripts/dist/netlify/index.js
"use strict";
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.token = exports.tokenCommand = exports.siteId = exports.siteIdCommand = exports.deploy = exports.deployCommand = exports.bandwidth = exports.bandwidthCommand = void 0;
var bandwidth_1 = require("./bandwidth");
Object.defineProperty(exports, "bandwidthCommand", { enumerable: true, get: function () { return __importDefault(bandwidth_1).default; } });
Object.defineProperty(exports, "bandwidth", { enumerable: true, get: function () { return bandwidth_1.execute; } });
var deploy_1 = require("./deploy");
Object.defineProperty(exports, "deployCommand", { enumerable: true, get: function () { return __importDefault(deploy_1).default; } });
Object.defineProperty(exports, "deploy", { enumerable: true, get: function () { return deploy_1.execute; } });
var site_id_1 = require("./site_id");
Object.defineProperty(exports, "siteIdCommand", { enumerable: true, get: function () { return __importDefault(site_id_1).default; } });
Object.defineProperty(exports, "siteId", { enumerable: true, get: function () { return site_id_1.execute; } });
var token_1 = require("./token");
Object.defineProperty(exports, "tokenCommand", { enumerable: true, get: function () { return __importDefault(token_1).default; } });
Object.defineProperty(exports, "token", { enumerable: true, get: function () { return token_1.execute; } });
