// @zendeskgarden/container-keyboardfocus/dist/index.cjs.js
/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var containerUtilities = require('@zendeskgarden/container-utilities');
var PropTypes = require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

var _excluded = ["tabIndex", "onBlur", "onFocus", "onMouseDown", "onPointerDown", "onTouchStart"];
function useKeyboardFocus() {
  var _useState = React.useState(false),
      keyboardFocused = _useState[0],
      setKeyboardFocused = _useState[1];
  var focusableTimeoutIdRef = React.useRef();
  var isKeyboardFocusableRef = React.useRef(true);
  React.useEffect(function () {
    return function () {
      clearTimeout(focusableTimeoutIdRef.current);
    };
  }, []);
  var onKeyboardFocusPointerDown = function onKeyboardFocusPointerDown() {
    isKeyboardFocusableRef.current = false;
    var timeoutId = setTimeout(function () {
      isKeyboardFocusableRef.current = true;
    }, 0);
    focusableTimeoutIdRef.current = Number(timeoutId);
  };
  var onKeyboardFocus = function onKeyboardFocus() {
    if (isKeyboardFocusableRef.current) {
      setKeyboardFocused(true);
    }
  };
  var onKeyboardFocusBlur = function onKeyboardFocusBlur() {
    setKeyboardFocused(false);
  };
  var getFocusProps = function getFocusProps(_temp) {
    var _ref = _temp === void 0 ? {} : _temp,
        _ref$tabIndex = _ref.tabIndex,
        tabIndex = _ref$tabIndex === void 0 ? 0 : _ref$tabIndex,
        onBlur = _ref.onBlur,
        onFocus = _ref.onFocus,
        onMouseDown = _ref.onMouseDown,
        onPointerDown = _ref.onPointerDown,
        onTouchStart = _ref.onTouchStart,
        props = _objectWithoutPropertiesLoose(_ref, _excluded);
    return _extends({
      tabIndex: tabIndex,
      onBlur: containerUtilities.composeEventHandlers(onBlur, onKeyboardFocusBlur),
      onFocus: containerUtilities.composeEventHandlers(onFocus, onKeyboardFocus),
      onMouseDown: containerUtilities.composeEventHandlers(onMouseDown, onKeyboardFocusPointerDown),
      onPointerDown: containerUtilities.composeEventHandlers(onPointerDown, onKeyboardFocusPointerDown),
      onTouchStart: containerUtilities.composeEventHandlers(onTouchStart, onKeyboardFocusPointerDown),
      'data-garden-container-id': 'containers.keyboardfocus',
      'data-garden-container-version': '0.4.14'
    }, props);
  };
  return {
    getFocusProps: getFocusProps,
    keyboardFocused: keyboardFocused
  };
}

var KeyboardFocusContainer = function KeyboardFocusContainer(_ref) {
  var children = _ref.children,
      _ref$render = _ref.render,
      render = _ref$render === void 0 ? children : _ref$render;
  return React__default["default"].createElement(React__default["default"].Fragment, null, render(useKeyboardFocus()));
};
KeyboardFocusContainer.propTypes = {
  children: PropTypes__default["default"].func,
  render: PropTypes__default["default"].func
};

exports.KeyboardFocusContainer = KeyboardFocusContainer;
exports.useKeyboardFocus = useKeyboardFocus;
